capability to reset necessary server attribs when they are unset

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
When admin unsets PBS Server attributes, they stop showing up on the pbs server but the server still uses their default values to operate normally.
This pull request changes the way these attributes are unset. With this change in place, unsetting some of the server attributes will result in resetting their values to default. These attributes will still show up on the server object even when they are unset.

#### Affected Platform(s)
All

#### Cause / Analysis / Design
* **[project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/924221441/Changes+to+how+default+attributes+are+set+on+PBS+server)***
*  **[pbspro community forum](http://community.pbspro.org/t/proposal-to-make-pbs-server-set-necessary-attributes-to-their-default-value-when-these-attributes-are-unset/1238/13)***


#### Solution Description
PBS server will reset the value of necessary attributes  -Table 3 on [design page
](https://pbspro.atlassian.net/wiki/spaces/PD/pages/924221441/Changes+to+how+default+attributes+are+set+on+PBS+server)

Some of the attributes are changing their default values (when unset) - Table 4 on [design page
](https://pbspro.atlassian.net/wiki/spaces/PD/pages/924221441/Changes+to+how+default+attributes+are+set+on+PBS+server)


#### Testing logs/output
[default_attr.txt](https://github.com/PBSPro/pbspro/files/2747613/default_attr.txt)
[selftest.txt](https://github.com/PBSPro/pbspro/files/2747614/selftest.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on src/server/pbsd_init.c:
```diff
@@ -317,6 +318,16 @@ init_server_attrs()
+
+	set_attr_svr(&(server.sv_attr[(int)SRV_ATR_query_others]), &svr_attr_def[(int) SRV_ATR_query_others], "TRUE");
+
+	set_attr_svr(&(server.sv_attr[(int)SRV_ATR_scheduling]), &svr_attr_def[(int) SRV_ATR_scheduling], "TRUE");
```
Use ATR_TRUE.  It's an include in pbs_internal.h.  Use ATR_TRUE/ATR_FALSE in other places in the code. 
 Also, did you test that the -a 0 option still works that this doesn't override it?

arungrover replied:
I tested it with -a0 option and it does set scheduling to false.

bhroam commented on src/server/req_manager.c:
```diff
@@ -1677,6 +1681,65 @@ mgr_server_unset(struct batch_request *preq)
+				char dflt_log_event[64];
+				sprintf(dflt_log_event, "%d", PBSEVENT_ERROR | PBSEVENT_SYSTEM | PBSEVENT_ADMIN | PBSEVENT_JOB |
+								PBSEVENT_JOB_USAGE | PBSEVENT_SECURITY | PBSEVENT_SCHED |
+								PBSEVENT_DEBUG | PBSEVENT_DEBUG2);
```
Since you create this bitfield twice, factor it out and put it in a macro.

bhroam commented on src/server/req_manager.c:
```diff
@@ -1677,6 +1681,65 @@ mgr_server_unset(struct batch_request *preq)
+					if (tm_list == NULL) {
+						reply_badattr(-1, bad_attr, plist, preq);
+					}
+					tm_list->al_link.ll_next->ll_struct = NULL;
```
Do you need to free anything first?

arungrover replied:
Since this is newly created list one 1 element, I don't think there is anything to free.

bhroam commented on test/tests/selftest/pbs_test_revert_to_defaults.py:
```diff
@@ -0,0 +1,82 @@
+        self.server.manager(MGR_CMD_SET, SERVER, modified_attr)
+        self.server.revert_to_defaults()
+
+        self.server.expect(SERVER, self.svr_dflt_attr, op=SET,
```
You don't need to say op=SET.  It is the default.

arungrover commented:
Thanks for reviewing @bhroam 
I have added another commit, I was not allocating enough memory for the values to be added to svrattrl structure. I've changed it now.

agrawalravi90 commented:
I have a question about the design: the table mentions that default_queue will now be unset (instead of being set to workq), but the text says that pbs_habitat will still set it:
Default_queue is an exception, Since pbs_habitat is responsible for creating the queue, setting the default_queue on the server object will also be with pbs_habitat.

So, will it be set or not?

agrawalravi90 commented on src/server/req_manager.c:
```diff
@@ -1649,13 +1651,16 @@ mgr_server_unset(struct batch_request *preq)
-				rc = mgr_unset_attr(dflt_scheduler->sch_attr, sched_attr_def, SCHED_ATR_LAST, tm_list,
-					-1, &bad_attr, (void *)dflt_scheduler, PARENT_TYPE_SCHED, INDIRECT_RES_CHECK);
+				/* when unset, set scheduler_iteration to 600 seconds */
+				sprintf(tm_list->al_value, "%d", 600);
```
Please use the macro PBS_SCHEDULE_CYCLE

agrawalravi90 commented on src/server/req_manager.c:
```diff
@@ -1677,6 +1682,68 @@ mgr_server_unset(struct batch_request *preq)
+			}
+			else if (strcasecmp(plist->al_name, ATTR_mailfrom) == 0)
+				set_attr_svr(&(server.sv_attr[(int)SRV_ATR_mailfrom]),
+					    &svr_attr_def[(int)SRV_ATR_mailfrom], "adm");
```
please use macros instead of string literals

agrawalravi90 commented on src/server/req_manager.c:
```diff
@@ -1677,6 +1682,68 @@ mgr_server_unset(struct batch_request *preq)
+			else if (strcasecmp(plist->al_name, ATTR_queryother) == 0)
+				set_attr_svr(&(server.sv_attr[(int)SRV_ATR_query_others]),
+					    &svr_attr_def[(int)SRV_ATR_query_others], "TRUE");
+			else if (strcasecmp(plist->al_name, ATTR_schediteration) == 0)
```
didn't we already handle sched iteration in the while block before in this function?  Also, should these 2 loops be combined?

agrawalravi90 commented on src/server/req_manager.c:
```diff
@@ -1649,13 +1651,16 @@ mgr_server_unset(struct batch_request *preq)
-					-1, &bad_attr, (void *)dflt_scheduler, PARENT_TYPE_SCHED, INDIRECT_RES_CHECK);
+				/* when unset, set scheduler_iteration to 600 seconds */
+				sprintf(tm_list->al_value, "%d", 600);
+				rc = mgr_set_attr(dflt_scheduler->sch_attr, sched_attr_def, SCHED_ATR_LAST, tm_list,
```
I'm don't know much about this code, but we will do an unset operation after this while loop is over, so does doing a set here make any difference?

agrawalravi90 commented on test/tests/functional/pbs_test_svr_dflt.py:
```diff
@@ -0,0 +1,81 @@
+        self.server.expect(SERVER, self.dflt_attr, attrop=PTL_AND,
+                           max_attempts=20)
+
+    def test_server_unset_dflt_attr_and_restart(self):
```
i kind of feel like you just need this one test, it  tests everything that test_server_unset_dflt_attr does

bhroam commented on src/server/req_manager.c:
```diff
@@ -1724,7 +1725,7 @@ mgr_server_unset(struct batch_request *preq)
 				if (plist->al_resc != NULL && strcasecmp(plist->al_resc, "ncpus") == 0) {
 					svrattrl *tm_list;
-					tm_list = attrlist_create(plist->al_name, "ncpus", 1);
+					tm_list = attrlist_create(plist->al_name, "ncpus", 2);
```
Why are you changing the resources_default.ncpus default from 1 to 2?  Shouldn't it be 1?

arungrover replied:
Well, 2 is not the resource value it is the amount of memory needed to store the value(which in our case is "1"). I'm changing it to 2 because it will need space for a string termination character as well.

mike0042 replied:
The svrattrl structure includes al_valln to keep track of the length of the value, which may contain multiple NULL bytes. Anything referencing svrattrl.al_value _should_ ensure it does not run past the end of the buffer. That said, it would be easy to mistakenly overrun the buffer if it is not NULL terminated. If we want to address this risk for all cases, we should probably do so in attrlist_create(). Just take whatever value size is passed and increment by one for the NULL terminator.

In the case of unsetting the default ncpus, we know we only need room for "1\0", but what if the value is changed again later? Do we create another attribute, or do we reuse the one created here? If we reuse this one, shouldn't there be enough space allocated to hold larger numbers?

I'll also note that strcpy() is used in attrlist_create() rather than rather than snprintf().

arungrover replied:
@mike0042 I like your suggestion of incrementing the length in attrlist_create function. 
In case of unsetting ncpus, if in future we plan to reset it back to a bigger numerical value, we will have to allocate enough space for it. Although, it seems unlikely that we will change the resources_default.ncpus to be a large value.
This attribute list is created only temporarily to make sure the mgr_set_attr makes a copy of it and it is immediately deleted afterward. I don't think we will have to reuse the attribute, it will always be created afresh.
For this code change, I'll increase the value in any case.

arungrover commented on pbspro.spec.in:
```diff
@@ -513,6 +513,7 @@ fi
 %exclude %{pbs_prefix}/lib*/python/pbs_bootcheck*
 %exclude %{pbs_prefix}/libexec/install_db
 %exclude %{pbs_prefix}/libexec/pbs_habitat
+%exclude %{pbs_prefix}/libexec/pbs_schema_upgrade
```
Yeah... this change is needed because when merged a PR of upgrading the DB schema, this line was added to pbspro.spec but it was missing from pbspro.spec.in.

arungrover commented on src/include/log.h:
```diff
@@ -139,6 +139,9 @@ extern int  setup_env(char *filename);
 #define PBSEVENT_FORCE		0x8000		/* set to force a messag      */
+#define SVR_LOG_DFLT		PBSEVENT_ERROR | PBSEVENT_SYSTEM | PBSEVENT_ADMIN | PBSEVENT_JOB \
+				| PBSEVENT_JOB_USAGE | PBSEVENT_SECURITY | PBSEVENT_SCHED \
+				| PBSEVENT_DEBUG | PBSEVENT_DEBUG2
```
Well, that is what the default was before. DEBUG2 was part of it. I think it should be ok

arungrover commented on src/server/pbsd_init.c:
```diff
@@ -317,6 +318,15 @@ init_server_attrs()
+	sprintf(dflt_log_event, "%d", SVR_LOG_DFLT);
+	set_attr_svr(&(server.sv_attr[(int)SRV_ATR_log_events]), &svr_attr_def[(int) SRV_ATR_log_events], dflt_log_event);
+
+	set_attr_svr(&(server.sv_attr[(int)SRV_ATR_mailfrom]), &svr_attr_def[(int) SRV_ATR_mailfrom], "adm");
```
I didn't know there exists a macro for this. I'll change it

arungrover commented on src/server/pbsd_init.c:
```diff
@@ -334,6 +344,15 @@ init_server_attrs()
+				ATR_VFLAG_DEFLT|ATR_VFLAG_SET|ATR_VFLAG_MODCACHE;
+			server.sv_attr[(int)SRV_ATR_resource_deflt].at_flags =
+				ATR_VFLAG_DEFLT|ATR_VFLAG_SET|ATR_VFLAG_MODCACHE;
+		}
```
I don't think we need to call action function here, it just check if the value of resource is +ve or not. Since we are setting the default value here, it is ok not to call it.

agrawalravi90 replied:
well that's what the action function does today, I think it's safer to call it, but I'll leave the decision to you.

arungrover replied:
I spoke to Bhroam about this, we both think that it is not needed to call action function unless it is doing more than just validating the value. In this case, since it is just checking for negative values, I'd rather leave it as is.

arungrover commented on src/server/req_manager.c:
```diff
@@ -1649,13 +1651,16 @@ mgr_server_unset(struct batch_request *preq)
 				svrattrl *tm_list;
-				tm_list = attrlist_create(plist->al_name, NULL, 0);
+				/* value is 600 so it is of size 4 including terminating character */
+				tm_list = attrlist_create(plist->al_name, NULL, 4);
```
I'll increase it to 8

arungrover commented on src/server/pbsd_init.c:
```diff
@@ -317,6 +318,15 @@ init_server_attrs()
 	attr_jobscript_max_size.at_type  |= ATR_TYPE_SIZE;  /* get_bytes_from_attr() is checking for at_type */
 	set_size(&attr_jobscript_max_size,&attrib,SET);
 
+	sprintf(dflt_log_event, "%d", SVR_LOG_DFLT);
```
I can use TOSTR but since SVR_LOG_DFLT is an ORed expression of other macros it will look similar to this in the string "0x01|0x02|0x04|0x08". 

This is why I need a string conversion

arungrover commented on src/server/pbsd_init.c:
```diff
@@ -253,6 +253,7 @@ init_server_attrs()
 	resource        *presc = NULL;
 	attribute	attrib;
 	int 		i = 0;
+	char		dflt_log_event[64];
```
I guess it is not needed to be that big as it just stores a number.  I'll reduce it to 8 bytes.

bhroam replied:
The log_event is stored in a long.  The max value for a signed long is 20 digits.  Toss in one more for the minus sign and one for the NULL byte, I'd make this 22 bytes at a minimum.

agrawalravi90 commented on src/server/pbsd_init.c:
```diff
@@ -253,7 +253,7 @@ init_server_attrs()
 	attribute	attrib;
 	int 		i = 0;
-	char		dflt_log_event[64];
+	char		dflt_log_event[8];
```
Please make this 22 like Bhroam suggested

arungrover replied:
ahh... I think I missed his comment. I'll change it

arungrover commented:
@agrawalravi90, @bhroam I've addressed your comments. Please have a look

arungrover commented:
I've added another commit. I needed to put default_chunk.ncpus back in PTL framework because scheduler does not reset default-chunk.ncpus now

bhroam commented:
I merged in the PR
